monai/data/image_writer.py

```diff
@@ -0,0 +1,395 @@
+    nib, has_nib = optional_import("nibabel")
+
+__all__ = ["ImageWriter", "ITKWriter"]
+
+
+class ImageWriter:
+    """
+    The class is a collection of utilities to write images to disk.
+
+    Main aspects to be considered are:
+
+        - dimensionality of the data array, arrangements of spatial dimensions and channel/time dimensions
+            - ``convert_to_channel_last()``
+        - metadata of the current affine and output affine, the data array should be converted accordingly
+            - ``get_meta_info()``
+            - ``resample_if_needed()``
+        - data type of the output image
+            - as part of ``resample_if_needed()``
+
+    Subclasses of this class should implement the backend-specific functions:
+
+        - ``set_data_array()`` to set the data array (input must be numpy array or torch tensor)
+            - this method sets the backend object's data part
+        - ``set_metadata()`` to set the metadata and output affine
+            - this method sets the metadata including affine handling and image resampling
+        - backend-specific data object
+            - ``create_backend_obj()``
+        - backend-specific writing function
+            - ``write()``
+
+    The primary usage is:
+
+    .. code-block:: python
+
+        writer = MyWriter()  # subclass of ImageWriter
+        writer.set_data_array(data_array)
+        writer.set_metadata(meta_dict).write(filename_or_obj)
+        writer.write(filename)
+
+    Create an image writer object based on ``data_array`` and ``metadata``.
+    Spatially it supports up to three dimensions (with the resampling step
+    supports both 2D and 3D).
+
+    When saving multiple time steps or multiple channels `data_array`, time
+    and/or modality axes should be the at the `channel_dim`. For example,
+    the shape of a 2D eight-class and channel_dim=0, the segmentation
+    probabilities to be saved could be `(8, 64, 64)`; in this case
+    ``data_array`` will be converted to `(64, 64, 1, 8)` (the third
+    dimension is reserved as a spatial dimension).
+
+    The ``metadata`` could optionally have the following keys:
+
+        - ``'original_affine'``: for data original affine, it will be the
+            affine of the output object, defaulting to an identity matrix.
+        - ``'affine'``: it should specify the current data affine, defaulting to an identity matrix.
+        - ``'spatial_shape'``: for data output spatial shape.
+
+    When ``metadata`` is specified and ``resample=True``, the saver will
+    try to resample data from the space defined by `"affine"` to the space
+    defined by `"original_affine"`, for more details, please refer to the
+    ``resample_if_needed`` method.
+    """
+
+    def __init__(self, **kwargs):
+        """the constructor supports adding new instance members."""
+        self.data_obj = None
+        for k, v in kwargs.items():
+            setattr(self, k, v)
```

Comment:
I didn't find the usage of these attributes.
And could you please help add doc-string to describe what's the expected values of this `kwargs` arg?

Thanks.

Reply:
After your initial comment, I've updated the docstring:
```
The constructor supports adding new instance members.
The current member in the base class is ``self.data_obj``, the subclasses can add more members,
so that necessary meta information can be stored in the object and shared among the class methods.
```